src/blockchain/electrum.rs

```diff
@@ -71,10 +71,139 @@ impl Blockchain for ElectrumBlockchain {
     fn setup<D: BatchDatabase, P: Progress>(
         &self,
         database: &mut D,
-        progress_update: P,
+        _progress_update: P,
     ) -> Result<(), Error> {
-        self.client
-            .electrum_like_setup(self.stop_gap, database, progress_update)
+        let mut request = script_sync::start(database, self.stop_gap)?;
+        let mut block_times = HashMap::<u32, u32>::new();
+        let mut txid_to_height = HashMap::<Txid, u32>::new();
+        let mut tx_cache = TxCache::new(database, &self.client);
+        let chunk_size = self.stop_gap;
+        // The electrum server has been inconsistent somehow in its responses during sync. For
+        // example, we do a batch request of transactions and the response contains less
+        // tranascations than in the request. This should never happen but we don't want to panic.
+        let electrum_goof = || Error::Generic("electrum server misbehaving".to_string());
+
+        let batch_update = loop {
+            request = match request {
+                Request::Script(script_req) => {
+                    let scripts = script_req.request().take(chunk_size);
+                    let txids_per_script: Vec<Vec<_>> = self
+                        .client
+                        .batch_script_get_history(scripts)
+                        .map_err(Error::Electrum)?
+                        .into_iter()
+                        .map(|txs| {
+                            txs.into_iter()
+                                .map(|tx| {
+                                    let tx_height = match tx.height {
+                                        none if none <= 0 => None,
+                                        height => {
+                                            txid_to_height.insert(tx.tx_hash, height as u32);
+                                            Some(height as u32)
+                                        }
+                                    };
+                                    (tx.tx_hash, tx_height)
+                                })
+                                .collect()
+                        })
+                        .collect();
+
+                    script_req.satisfy(txids_per_script)?
+                }
+
+                Request::Conftime(conftime_req) => {
```

Comment:
The blockchain implementation should not rely on the implementation details of `script_sync`. It has changed during the course of this PR and may change again.  I think I don't keep to that in esplora because I expect the `ScriptReq` to come first. I'll try and fix that.

Reply:
On second thought, I think it's not worth doing in this PR since those branches won't be tested yet. When we do partial syncs where tx conf times are requested without having been found by ` script_req` first we can fix this.

Reply:
Yes makes sense. Can be done on a future PR.

corollary question, given the structure of `tx_needed` and `tx_missing_conftime` in sync state https://github.com/bitcoindevkit/bdk/blob/77266c4a9180df82bf98523df1a0bc7681bb1cc0/src/blockchain/script_sync.rs#L277-L281 

It seems to me that the logical order of requests is always `Script -> Tx -> Conftimes`. Because `Conftime` will need tx details which we can't fetch until we do a `Tx` request on the ID, and `Script` is what we always start with. Is this correct thinking? Or the order can change in different blockchain backend?

Reply:
With the current internals yes that will be the order but you could easily change the internals to do it in a different way. The point is the blockchain should no care (ideally).